graphula-core/src/Graphula.hs

```diff
@@ -429,85 +467,234 @@ class HasDependencies a where
+  genEntityKey = Just <$> arbitrary
+
+type family Quote t where
+  Quote t = 'Text "‘" ':<>: t ':<>: 'Text "’"
+
+-- | Explicit instance for @'SourceExternal'@ to give an actionable error message
+--
+-- Rendered:
+--
+-- @
+-- Cannot generate a value of type ‘X’ using ‘node’ since
+--
+--   instance HasDependencies X where
+--     type KeySource X = 'SourceExternal
+--
+-- Possible fixes include:
+-- • Use ‘nodeKeyed’ instead of ‘node’
+-- • Change ‘KeySource X’ to 'SourceDefault or 'SourceArbitrary
+-- @
+--
+instance TypeError
+  ( 'Text "Cannot generate a value of type "
+    ':<>: Quote ('ShowType a)
+    ':<>: 'Text " using "
+    ':<>: Quote ('Text "node")
+    ':<>: 'Text " since"
+    ':$$: 'Text ""
+    ':$$: 'Text "  instance HasDependencies "
+    ':<>: 'ShowType a
+    ':<>: 'Text " where"
+    ':$$: 'Text "    "
+    ':<>: 'Text "type KeySource "
+    ':<>: 'ShowType a
+    ':<>: 'Text  " = "
+    ':<>: 'ShowType 'SourceExternal
+    ':$$: 'Text ""
+    ':$$: 'Text "Possible fixes include:"
+    ':$$: 'Text "• Use "
+    ':<>: Quote ('Text "nodeKeyed")
+    ':<>: 'Text " instead of "
+    ':<>: Quote ('Text "node")
+    ':$$: 'Text "• Change "
+    ':<>: Quote ('Text "KeySource " ':<>: 'ShowType a)
+    ':<>: 'Text " to "
+    ':<>: 'Text "'SourceDefault"
+    ':<>: 'Text " or "
+    ':<>: 'Text "'SourceArbitrary"
+  ) => GenEntityKey 'SourceExternal a where
+  type KeyConstraint 'SourceExternal a = NoConstraint a
+  genEntityKey = error "unreachable"
+
+data GenerationFailure
+  = GenerationFailureMaxAttemptsToConstrain TypeRep
+  -- ^ Could not satisfy constraints defined using @'ensure'@
+  | GenerationFailureMaxAttemptsToInsert TypeRep
+  -- ^ Could not satisfy database constraints on insert
   deriving (Show, Typeable, Eq)
 
 instance Exception GenerationFailure
 
-type GraphulaNode m a =
-  ( Generate m a
-  , HasDependencies a
-  , Logging m a
-  , PersistEntityBackend a ~ SqlBackend
-  , PersistEntity a
-  , Typeable a
-  , EntityKeyGen a
-  )
+type GraphulaNode m a
+  = ( Generate m a
+    , HasDependencies a
+    , Logging m a
+    , PersistEntityBackend a ~ SqlBackend
+    , PersistEntity a
+    , Typeable a
+    )
 
 {-|
-  Generate and edit a value with data dependencies. This leverages
+  Generate a value with data dependencies. This leverages
   'HasDependencies' to insert the specified data in the generated value. All
-  dependency data is inserted after editing operations.
+  dependency data is inserted after any editing operations.
 
-  > nodeEdit @Dog (ownerId, veterinarianId) $ \dog ->
+  > node @Dog (ownerId, veterinarianId) mempty
+  > node @Dog (ownerId, veterinarianId) $ edit $ \dog ->
   >   dog {name = "fido"}
+
+  A value that has an externally managed key must use @'nodeKeyed'@ instead.
 -}
-nodeEditWith
-  :: forall a m
-   . GraphulaContext m '[a]
+node
+  :: forall a s m
+   . ( GraphulaContext m '[a]
+     , KeySource a ~ s
+     , GenEntityKey s a
```

Comment:
So if I call this function with a `SouceExternal` type, then I will get a type error from the instance resolution?

Reply:
Yep

```plaintext
/home/ubuntu/code/graphula/graphula-core/test/README.lhs:207:10: error:
    • Cannot generate a value of type ‘E’ using ‘node’ since
            
        instance HasDependencies E where
          type KeySource E = 'SourceExternal
            
      Possible fixes include:
      • Use ‘nodeKeyed’ instead of ‘node’
      • Change ‘KeySource E’ to 'SourceDefault or 'SourceArbitrary
    • In a stmt of a 'do' block: _ <- node @E () mempty
      In the second argument of ‘($)’, namely
        ‘do Entity aId _ <- node @A () mempty
            liftIO $ putStrLn "A"
            Entity bId b <- node @B (only aId) mempty
            liftIO $ putStrLn "B"
            ....’
      In the expression:
        runGraphulaT runDB
          $ do Entity aId _ <- node @A () mempty
               liftIO $ putStrLn "A"
               Entity bId b <- node @B (only aId) mempty
               liftIO $ putStrLn "B"
               ....
    |       
207 |     _ <- node @E () mempty
    |          ^^^^^^^^^^^^^^^^^
```

That's what the big mess of `TypeError` constraints is doing up above

Reply:
Beautiful